base/src/commonMain/kotlin/collections/ConcurrentMutableList.kt

```diff
@@ -0,0 +1,69 @@
+    distributed under the License is distributed on an "AS IS" BASIS,
+    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+    See the License for the specific language governing permissions and
+    limitations under the License.
+
+ */
+
+package com.splendo.kaluga.base.collections
+
+import kotlinx.atomicfu.locks.reentrantLock
+import kotlinx.atomicfu.locks.withLock
+
+class ConcurrentMutableList<E> internal constructor(private val internal: MutableList<E> = mutableListOf()) : MutableList<E> {
+    private val lock = reentrantLock()
+
+    override val size: Int get() = synchronized { size }
+    override fun add(element: E): Boolean = synchronized { add(element) }
+    override fun add(index: Int, element: E) = synchronized { add(index, element) }
+    override fun addAll(elements: Collection<E>): Boolean = synchronized { addAll(elements) }
+    override fun addAll(index: Int, elements: Collection<E>): Boolean = synchronized { addAll(index, elements) }
+    override fun get(index: Int): E = synchronized { get(index) }
+    override fun set(index: Int, element: E): E = synchronized { set(index, element) }
+    override fun indexOf(element: E): Int = synchronized { indexOf(element) }
+    override fun lastIndexOf(element: E): Int = synchronized { lastIndexOf(element) }
+    override fun contains(element: E): Boolean = synchronized { contains(element) }
+    override fun containsAll(elements: Collection<E>): Boolean = synchronized { containsAll(elements) }
+    override fun iterator(): ConcurrentMutableIterator<E> = synchronized { ConcurrentMutableIterator(iterator()) }
+    override fun listIterator(): ConcurrentMutableListIterator<E> = synchronized { ConcurrentMutableListIterator(listIterator()) }
+    override fun listIterator(index: Int): ConcurrentMutableListIterator<E> = synchronized { ConcurrentMutableListIterator(listIterator(index)) }
+    override fun remove(element: E): Boolean = synchronized { remove(element) }
+    override fun removeAt(index: Int): E = synchronized { removeAt(index) }
+    override fun removeAll(elements: Collection<E>): Boolean = synchronized { removeAll(elements.toSet()) }
+    override fun retainAll(elements: Collection<E>): Boolean = synchronized { retainAll(elements.toSet()) }
+    override fun clear() = synchronized { clear() }
+    override fun subList(fromIndex: Int, toIndex: Int): ConcurrentMutableList<E> = synchronized { ConcurrentMutableList(subList(fromIndex, toIndex)) }
+    override fun isEmpty(): Boolean = synchronized { isEmpty() }
+    override fun equals(other: Any?): Boolean = when (other) {
+        is ConcurrentMutableSet<*> -> {
+            synchronized {
+                val inner = this
+                other.synchronized {
+                    inner == this
+                }
+            }
+        }
+        is MutableSet<*> -> synchronized {
+            this == other
+        }
+        else -> false
```

Comment:
In case of both a lists, should you compare its content (like build-in list does)?